### PR TITLE
Dev/70/alex hierarchy

### PR DESCRIPTION
**Pull request by alexcouper**
This works out the hierarchy of branches and exposes them correctly in the API.

`get_feature_info` needs to be written in the github provider.

### CODE REVIEW

**Comment by alexcouper** [comment]
To try it out go to `http://localhost:8000/features/github/70` in your browser.

In time we'll want to tweak it so that we don't repeat the inclusion of children in the JSON data.

File: deploystream/lib/hierarchy.py [hunk 1]
```diff
@@ -1,31 +1,10 @@
+from collections import defaultdict
 import re
 
 
-def create_single_regex(feature_id, hierarchical_regexes):
+def match_with_geneology(feature_id, branches, hierarchical_regexes):
```
Comment (txels) [inline deploystream/lib/hierarchy.py]: genealogy? http://en.wikipedia.org/wiki/Genealogy

File: deploystream/apps/feature/models.py [hunk 2]
```diff
@@ -38,16 +41,19 @@ def __init__(self, provider, project, id, title,
...
         self.trees = []
 
-    def create_hierarchy_trees(self):
-        "Create hierarchy trees - one for each repo."
-        pass
-
     def add_branch(self, branch):
         assert isinstance(branch, Branch)
-        self.branches.append(branch)
+        self.branches[branch.repo_name][branch.branch_name] = branch
```
Comment (txels) [inline deploystream/apps/feature/models.py]: This nested-dict data structure is quite difficult to handle in the front-end. Lists of dicts would be much easier to handle, as they directly translate in Javascript into iterable lists of objects.

See my other PR for more details https://github.com/pretenders/deploystream/pull/83

File: deploystream/apps/feature/models.py [hunk 3]
```diff
@@ -56,15 +62,13 @@ class Branch(object):
...
-        ``branch_name``   - The name of the branch.
-        ``latest_commit`` - The head commmit, or latest revision in this
-                            branch.
-        ``level``         - The numerical level that this branch falls in the
-                            hierarchy for the feature - where 0 is the highest
-                            level.
-        ``provider``      - The provider instance that found this branch
-                            information.
+        ``repo_name``           - The repository that this branch is found in.
+        ``branch_name``         - The name of the branch.
```
Comment (txels) [inline deploystream/apps/feature/models.py]: As this is a `Branch`, `name` would suffice.

File: deploystream/apps/feature/models.py [hunk 4]
```diff
@@ -56,15 +62,13 @@ class Branch(object):
...
+        ``latest_commit``       - The head commmit, or latest revision in this
```
Comment (txels) [inline deploystream/apps/feature/models.py]: `commit_id`?

File: deploystream/apps/feature/models.py [hunk 5]
```diff
@@ -56,15 +62,13 @@ class Branch(object):
...
-        ``repo_name``     - The repository that this branch is found in.
-        ``branch_name``   - The name of the branch.
-        ``latest_commit`` - The head commmit, or latest revision in this
-                            branch.
-        ``level``         - The numerical level that this branch falls in the
-                            hierarchy for the feature - where 0 is the highest
-                            level.
-        ``provider``      - The provider instance that found this branch
-                            information.
+        ``repo_name``           - The repository that this branch is found in.
```
Comment (txels) [inline deploystream/apps/feature/models.py]: Trying to review these names from the point of view of the resulting JSON data...

`repository`?

**Comment by alexcouper** [comment]
@txels I've made those changes - although I haven't made the flat storage change as I wasn't sure if what you've done in your PR solved that issue for you?

**Comment by txels** [comment]
The change in my PR actually does not "fix" the data structure mismatch, just adds more data with a different structure and the JS code ignores the rest. The proper fix will be to settle on a single structure that serves all use cases.

**Pull request merged**